internal/common/types/component/constructor.go

```diff
@@ -177,60 +174,63 @@ func (c *Constructor) AddImageAsResource(imageInfos []*image.ImageInfo) {
 	}
 }
 
-func (c *Constructor) AddRawManifestResource(manifestPath string) {
+func (c *Constructor) AddFileAsDirResource(resourceName, filePath string) error {
+	dir, err := getAbsPath(filepath.Dir(filePath))
+	if err != nil {
+		return err
+	}
+
 	c.Components[0].Resources = append(c.Components[0].Resources, Resource{
-		Name:    common.RawManifestResourceName,
+		Name:    resourceName,
 		Type:    DirectoryTreeResourceType,
 		Version: c.Components[0].Version,
 		Input: &Input{
 			Type:         DirectoryInputType,
-			Path:         filepath.Dir(manifestPath),
+			Path:         dir,
 			Compress:     true,
-			IncludeFiles: []string{filepath.Base(manifestPath)},
+			IncludeFiles: []string{filepath.Base(filePath)},
 		},
 	})
+	return nil
+}
+
+func getAbsPath(filePath string) (string, error) {
+	if !filepath.IsAbs(filePath) {
+		absPath, err := filepath.Abs(filePath)
+		if err != nil {
+			return "", fmt.Errorf("failed to get absolute path for %s: %w", filePath, err)
+		}
+		filePath = absPath
+	}
+	return filePath, nil
 }
 
-func (c *Constructor) AddDefaultCRResource(defaultCRPath string) {
+func (c *Constructor) AddBinaryDataAsFileResource(resourceName string, data []byte) {
 	c.Components[0].Resources = append(c.Components[0].Resources, Resource{
-		Name:    common.DefaultCRResourceName,
-		Type:    DirectoryTreeResourceType,
+		Name:    resourceName,
+		Type:    PlainTextResourceType,
 		Version: c.Components[0].Version,
 		Input: &Input{
-			Type:         DirectoryInputType,
-			Path:         filepath.Dir(defaultCRPath),
-			Compress:     true,
-			IncludeFiles: []string{filepath.Base(defaultCRPath)},
+			Type: BinaryResourceInput,
+			Data: base64.StdEncoding.EncodeToString(data),
 		},
 	})
 }
 
-func (c *Constructor) AddMetadataResource(moduleConfig *contentprovider.ModuleConfig) error {
-	yamlData, err := resources.GenerateMetadataYaml(moduleConfig)
+func (c *Constructor) AddFileResource(resourceName, filePath string) error {
```

Comment:
I think you can design this as an interface method, and internally use factory pattern to switch between different approaches. It simplifies the logic to the client, and also does not necessarily expose those `AddXXX` externally.

Something like this:
```
func (c *Constructor) AddFileResource(resourceName, filePath string) error {
	switch resourceName {
	case common.RawManifestResourceName, common.DefaultCRResourceName:
		return c.addFileAsDirResource(resourceName, filePath)
	case common.ModuleTemplateResourceName:
		return c.addFileResource(resourceName, filePath)
	default:
		return fmt.Errorf("unknown resource name: %s", resourceName)
	}
}
``